src/heartbeat.ts

```diff
@@ -0,0 +1,130 @@
+        this._register(vscode.debug.onDidStartDebugSession(this.updateLastActivitiy, this));
+        this._register(vscode.debug.onDidReceiveDebugSessionCustomEvent(this.updateLastActivitiy, this));
+        this._register(vscode.debug.onDidTerminateDebugSession(this.updateLastActivitiy, this));
+        this._register(vscode.debug.onDidChangeBreakpoints(this.updateLastActivitiy, this));
+        this._register(vscode.extensions.onDidChange(this.updateLastActivitiy, this));
+        this._register(vscode.languages.onDidChangeDiagnostics(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidStartTask(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidStartTaskProcess(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidEndTask(this.updateLastActivitiy, this));
+        this._register(vscode.tasks.onDidEndTaskProcess(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeWorkspaceFolders(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidOpenTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCloseTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidSaveTextDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidSaveNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidOpenNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCloseNotebookDocument(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillCreateFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidCreateFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillDeleteFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidDeleteFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onWillRenameFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidRenameFiles(this.updateLastActivitiy, this));
+        this._register(vscode.workspace.onDidChangeConfiguration(this.updateLastActivitiy, this));
+        this._register(vscode.languages.registerHoverProvider('*', {
+            provideHover: () => {
+                this.updateLastActivitiy();
+                return null;
+            }
+        }));
+
+        this.logger.trace(`Heartbeat manager for workspace ${workspaceId} (${instanceId}) - ${gitpodHost} started`);
+
+        // Start heatbeating interval
+        this.sendHeartBeat();
+        this.heartBeatHandle = setInterval(() => {
```

Comment:
interesting that this logic was changed in supervisor frontend to be less aggressive: https://github.com/gitpod-io/gitpod/blob/f157a029aeb9bcd37a29082dd13dc86a8c2d09e2/components/supervisor/frontend/src/ide/heart-beat.ts#L62-L71

I think we should align to reduce load as well. See https://github.com/gitpod-io/gitpod/pull/5613